### PR TITLE
Pregnancy Progress rework

### PR DESCRIPTION
**Pull request by brrritssocold**
This PR is ongoing work in response to #322. Phew! That was a lot of work.

I opted for a lookup approach as is scales better than endless `if` blocks. A scene can be registered with Pregnancy progression, specifying Mother, Father and what scene to call.

In addition it can be extended to nest maps, so Mother - > Father -> Scene lookups can be performed.

- Characterisation tests for Birth
- Extracted code into methods
- Code cleanup (spacing, curly brackets) for extracted code
- Interface for Vaginal pregnancy
- Use map / associative array / dictionary for scene lookup
- Extract pregnancy content into own files where applicable

TODO:
 - Extract anal pregnancy
 - Extend `PregnancyProgression` where player is not the mother, using nested maps
   - e.g. Cotton impregnated by player, SandWitch impregnated by player, etc.
   - Allow for any pregnancy combination (NPC/NPC, NPC/Monster, Player/Player [Sheila scene, see #261], etc.)
 - How to deal with non-standard pregnancies (Benoit), Pre / Post hooks?
 - If possible extract default progression code ("Your belly has gotten larger", etc.)

### CODE REVIEW

File: classes/classes/Scenes/Monsters/pregnancies/PlayerBasiliskPregnancy.as [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		/**
+		 * Create a new Basilisk pregnancy for the player. Registers pregnancy for Basilisk.
+		 * @param	pregnancyProgression instance used for registering pregnancy scenes
+		 * @param	output instance for gui output
+		 */
+		public function PlayerBasiliskPregnancy(pregnancyProgression:PregnancyProgression, output:GuiOutput) 
+		{
+			// needed as an instance variable for refactor test code (detectVaginalBirth)
+			this.pregnancyProgression = pregnancyProgression;
+			this.output = output;
```
Comment (Stadler76) [inline classes/classes/Scenes/Monsters/pregnancies/PlayerBasiliskPregnancy.as]: I'm not a fan of passing references around the codebase through constructors, like this.

As an example I've set up `Output` as a singleton, so
```as3
output = Output.init();
```
Should do without the need of the `output` param.

For the first param maybe something, like
```as3
		public function PlayerBasiliskPregnancy(pregnancyProgression:PregnancyProgression = null) 
		{
			// needed as an instance variable for refactor test code (detectVaginalBirth)
			this.pregnancyProgression = pregnancyProgression || PregnancyProgression.init();
```
Or something like that could work.

File: classes/classes/Scenes/Monsters/pregnancies/PlayerBunnyPregnancy.as [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		 * @param	output instance for gui output
+		 */
+		public function PlayerBunnyPregnancy(pregnancyProgression:PregnancyProgression, output:GuiOutput, mutations:Mutations)
+		{
+			this.pregnancyProgression = pregnancyProgression;
+			this.output = output;
+			this.mutations = mutations;
+			
+			pregnancyProgression.registerVaginalPregnancyScene(PregnancyStore.PREGNANCY_PLAYER, PregnancyStore.PREGNANCY_BUNNY, this);
+		}
```
Comment (Stadler76) [inline classes/classes/Scenes/Monsters/pregnancies/PlayerBunnyPregnancy.as]: This should do as well:
```as3
		public function PlayerBunnyPregnancy(pregnancyProgression:PregnancyProgression)
		{
			this.pregnancyProgression = pregnancyProgression;
			output = Output.init();
			mutations = Mutations.init();
			
			pregnancyProgression.registerVaginalPregnancyScene(PregnancyStore.PREGNANCY_PLAYER, PregnancyStore.PREGNANCY_BUNNY, this);
		}
```

Well, just a suggestion, no 'must do'.

Comment (brrritssocold) [reply classes/classes/Scenes/Monsters/pregnancies/PlayerBunnyPregnancy.as]: Unfortunately, the example you provided `output = Output.init();` is a textbook case of Singleton abuse.

Singletons should be used when only one instance of a class should exist. Unfortunately they are often misused for global state and 'easy' access, as you just demonstrated in the example.
This leads to classes lying about their dependencies.

Related is the rampant inheritance abuse in the form of `BaseContent`.
The reason for instance why the call `new Creature();`
will fail without 
```actionscript
		[BeforeClass]
		public static function setUpClass():void {
			kGAMECLASS = new CoC(StageLocator.stage);
		}
```

Further reading:
- [Singletons are Pathological Liars, Miško Hevery](http://misko.hevery.com/2008/08/17/singletons-are-pathological-liars/)
- [When is Singleton appropriate?, Stack Exchange](https://softwareengineering.stackexchange.com/questions/252/when-is-singleton-appropriate)

Comment (Stadler76) [reply classes/classes/Scenes/Monsters/pregnancies/PlayerBunnyPregnancy.as]: @brrritssocold wrote:
> Singletons should be used when only one instance of a class should exist.

Actually that was the intention of making those two into a singleton.

**Pull request merged**